File: src/main/java/com/demigodsrpg/chitchat/ChatListener.java
package com.demigodsrpg.chitchat;

import io.papermc.paper.event.player.AsyncChatEvent;
import net.kyori.adventure.text.Component;
import net.kyori.adventure.text.event.ClickEvent;
import net.kyori.adventure.text.format.NamedTextColor;
import net.kyori.adventure.text.format.TextDecoration;
import net.kyori.adventure.text.serializer.plain.PlainTextComponentSerializer;
import org.bukkit.Bukkit;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.EventPriority;
import org.bukkit.event.Listener;
import org.bukkit.event.player.PlayerCommandPreprocessEvent;
import org.bukkit.event.player.PlayerQuitEvent;

public class ChatListener implements Listener {

    private final Chitchat INST;
    private final Component DELETE_CROSS_BASE = Component.textOfChildren(
            Component.text("[", NamedTextColor.DARK_GRAY),
            Component.text("X", NamedTextColor.DARK_RED, TextDecoration.BOLD),
            Component.text("]", NamedTextColor.DARK_GRAY)
    ).hoverEvent(Component.text("Delete?", NamedTextColor.RED));

    public ChatListener(Chitchat inst) {
        INST = inst;
    }

    // -- BUKKIT CHAT LISTENER -- //

    @EventHandler(priority = EventPriority.LOW, ignoreCancelled = true)
    public void onChat(AsyncChatEvent chat) {
        if (INST.getMuteMap().containsKey(chat.getPlayer().getUniqueId().toString())) {
            chat.setCancelled(true);
        }
    }

    @EventHandler(priority = EventPriority.HIGHEST, ignoreCancelled = true)
    public void onFinalChat(AsyncChatEvent chat) {
        final Component formattedMessage = INST.FORMAT.getFormattedMessage(chat.getPlayer(), chat.originalMessage());
        chat.renderer((source, sourceDisplayName, message, viewer) -> {
            Component finalMessage = formattedMessage;

            if(viewer == source || (viewer instanceof Player player && player.hasPermission("chitchat.delete"))) {
                Component deleteCross = DELETE_CROSS_BASE.
                        clickEvent(ClickEvent.callback(audience -> Bukkit.getServer().deleteMessage(chat.signedMessage())));
                finalMessage =  finalMessage.appendSpace().append(deleteCross);
            }

            return finalMessage;
        });
    }

    @EventHandler(priority = EventPriority.HIGHEST, ignoreCancelled = true)
    public void onPreprocessCommand(PlayerCommandPreprocessEvent command) {
        Player player = command.getPlayer();
        String[] commandMsg = command.getMessage().split("\\s+");

        // Muted commands
        if (INST.getMuteMap().containsKey(player.getUniqueId().toString())) {
            if (Chitchat.getInst().MUTED_COMMANDS.contains(commandMsg[0].toLowerCase().substring(1))) {
                command.setCancelled(true);
                player.sendMessage(Component.text("I'm sorry " + player.getName() + ", I'm afraid I can't do that.", NamedTextColor.RED));
            }
        }

        // /me <message>
        else if (Chitchat.getInst().OVERRIDE_ME && commandMsg.length > 1 && commandMsg[0].equals("/me")) {
            command.setCancelled(true);
            if (Chitchat.getInst().MUTED_COMMANDS.contains("me") && INST.getMuteMap().containsKey(player.
                    getUniqueId().toString())) {
                player.sendMessage(Component.text("I'm sorry " + player.getName() + ", I'm afraid I can't do that.", NamedTextColor.RED));
            } else {
                Component message = Component.text(PlainTextComponentSerializer.plainText().serialize(player.displayName()) + " " + command.getMessage().substring(4)).decorate(TextDecoration.ITALIC);
                Bukkit.getServer().sendMessage(message);
            }
        }
    }

    @EventHandler(priority = EventPriority.HIGHEST)
    public void onQuit(PlayerQuitEvent event) {
        // Remove all useless reply data
        String playerName = event.getPlayer().getName();
        INST.getReplyMap().remove(playerName);
        if (INST.getReplyMap().containsValue(playerName)) {
            INST.getReplyMap().entrySet().stream().
                    filter(entry -> entry.getValue().equals(playerName)).
                    forEach(entry -> INST.getReplyMap().remove(entry.getKey()));
        }
    }
}

File: src/main/kotlin/com/demigodsrpg/chitchat/ChatListener.kt
package com.demigodsrpg.chitchat

import com.demigodsrpg.chitchat.tag.ChatScope
import net.md_5.bungee.api.chat.TextComponent
import org.bukkit.ChatColor
import org.bukkit.event.EventHandler
import org.bukkit.event.EventPriority
import org.bukkit.event.Listener
import org.bukkit.event.player.AsyncPlayerChatEvent
import org.bukkit.event.player.PlayerCommandPreprocessEvent
import org.bukkit.event.player.PlayerQuitEvent

class ChatListener(private val INST: Chitchat) : Listener {

    // -- BUKKIT CHAT LISTENER -- //

    @EventHandler(priority = EventPriority.LOW, ignoreCancelled = true)
    fun onChat(chat: AsyncPlayerChatEvent) {
        if (INST.MUTE_MAP.keys.contains(chat.player.uniqueId.toString())) {
            chat.isCancelled = true
        }
    }

    @EventHandler(priority = EventPriority.HIGHEST, ignoreCancelled = true)
    fun onFinalChat(chat: AsyncPlayerChatEvent) {
        Chitchat.sendMessage(INST.FORMAT.getFormattedMessage(chat.player, ChatScope.LOCAL, chat.message),
                chat.recipients)
        if (Chitchat.INST.USE_REDIS && !INST.FORMAT.shouldCancelRedis(chat.player)) {
            RChitchat.REDIS_CHAT.publishAsync(RChitchat.serverId + "$" +
                    INST.FORMAT.getSerializedMessage(chat.player, ChatScope.CHANNEL, chat.message))
        }
        chat.recipients.clear()
    }

    @EventHandler(priority = EventPriority.HIGHEST, ignoreCancelled = true)
    fun onPreprocessCommand(command: PlayerCommandPreprocessEvent) {
        val player = command.player
        val commandMsg = command.message.split("\\s+".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()

        // Muted commands
        if (INST.MUTE_MAP.keys.contains(player.uniqueId.toString())) {
            if (Chitchat.INST.MUTED_COMMANDS.contains(commandMsg[0].toLowerCase().substring(1))) {
                command.isCancelled = true
                player.sendMessage(ChatColor.RED.toString() + "I'm sorry " + player.name +
                        ", I'm afraid I can't do that.")
            }
        } else if (Chitchat.INST.OVERRIDE_ME && commandMsg.size > 1 && commandMsg[0] == "/me") { // /me <message>
            command.isCancelled = true
            if (Chitchat.INST.MUTED_COMMANDS.contains("me") && INST.MUTE_MAP.keys.
                    contains(player.uniqueId.toString())) {
                player.sendMessage(ChatColor.RED.toString() + "I'm sorry " + player.name +
                        ", I'm afraid I can't do that.")
            } else {
                var message = command.message.substring(1)
                message = ChatColor.ITALIC.toString() + ChatColor.stripColor(player.displayName + " " +
                        message.substring(3))
                Chitchat.sendMessage(TextComponent(message))
            }
        }
    }

    @EventHandler(priority = EventPriority.HIGHEST)
    fun onQuit(event: PlayerQuitEvent) {
        // Remove all useless reply data
        val playerName = event.player.name
        INST.REPLY_MAP.remove(playerName)
        if (INST.REPLY_MAP.containsValue(playerName)) {
            INST.REPLY_MAP.entries.stream().filter({ entry -> entry.value == playerName }).
                    forEach { entry -> INST.REPLY_MAP.remove(entry.key) }
        }
    }
}
